File: src-tauri/src/domain/work_item.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

pub const WORK_ITEM_SCHEMA_VERSION: u32 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkItemKind {
    Task,
    Obligation,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ObligationMode {
    OneOff,
    Ongoing,
    Recurring,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkItemStatus {
    Planned,
    Ready,
    InProgress,
    Waiting,
    Blocked,
    Review,
    Completed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkItemPriority {
    Low,
    Normal,
    High,
    Critical,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActorType {
    User,
    System,
    Integration,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PartyReference {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub display_name: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkItemEvent {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sequence: Option<i64>,
    pub at: String,
    pub actor_type: ActorType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actor_name: Option<String>,
    pub action: String,
    pub summary: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub changes: Vec<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkItemContextEntry {
    pub id: String,
    #[serde(rename = "type")]
    pub entry_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    pub created_at: String,
    pub actor_type: ActorType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actor_name: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExternalReference {
    #[serde(rename = "type")]
    pub reference_type: String,
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub url: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkItem {
    pub schema_version: u32,
    pub id: String,
    pub kind: WorkItemKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub obligation_mode: Option<ObligationMode>,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expected_outcome: Option<String>,
    pub status: WorkItemStatus,
    pub priority: WorkItemPriority,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub owner: Option<PartyReference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accountable_to: Option<PartyReference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub area_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_review_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blocker_reason: Option<String>,
    pub source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub planning_bucket: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub estimate_minutes: Option<i64>,
    #[serde(default)]
    pub external_references: Vec<ExternalReference>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub context: Vec<WorkItemContextEntry>,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub archived_at: Option<String>,
    #[serde(default)]
    pub history: Vec<WorkItemEvent>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, Value>,
}

impl WorkItem {
    pub fn validate(&self) -> Result<(), String> {
        if self.schema_version != WORK_ITEM_SCHEMA_VERSION {
            return Err(format!(
                "Unsupported work item schema version {}.",
                self.schema_version
            ));
        }
        if self.id.trim().is_empty() {
            return Err("Work item id must not be empty.".to_string());
        }
        if self.title.trim().is_empty() {
            return Err("Work item title must not be empty.".to_string());
        }
        if self.revision < 1 {
            return Err("Work item revision must be at least 1.".to_string());
        }
        if self.kind == WorkItemKind::Task && self.obligation_mode.is_some() {
            return Err("A task cannot define obligationMode.".to_string());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serializes_provider_neutral_camel_case_contract() {
        let item = WorkItem {
            schema_version: WORK_ITEM_SCHEMA_VERSION,
            id: "work-1".to_string(),
            kind: WorkItemKind::Task,
            obligation_mode: None,
            title: "Confirm renewal".to_string(),
            description: None,
            expected_outcome: None,
            status: WorkItemStatus::InProgress,
            priority: WorkItemPriority::Normal,
            owner: None,
            accountable_to: None,
            area_id: None,
            parent_id: None,
            start_at: None,
            due_at: None,
            completed_at: None,
            next_review_at: None,
            blocker_reason: None,
            source: "manual".to_string(),
            source_url: None,
            planning_bucket: None,
            estimate_minutes: None,
            external_references: vec![],
            tags: vec![],
            context: vec![],
            created_at: "2026-07-29T10:00:00Z".to_string(),
            updated_at: "2026-07-29T10:00:00Z".to_string(),
            revision: 1,
            archived_at: None,
            history: vec![],
            metadata: HashMap::new(),
        };

        item.validate().unwrap();
        let value = serde_json::to_value(item).unwrap();
        assert_eq!(value["schemaVersion"], 1);
        assert_eq!(value["status"], "in_progress");
        assert!(value.get("obligationMode").is_none());
    }
}

File: src/domain/workItem.ts
import type { PlanningBucket, Task, TaskObligationKind, TaskStatus } from '../types';

export const WORK_ITEM_SCHEMA_VERSION = 1 as const;

export type WorkItemKind = 'task' | 'obligation';
export type ObligationMode = 'one_off' | 'ongoing' | 'recurring';
export type WorkItemStatus =
  | 'planned'
  | 'ready'
  | 'in_progress'
  | 'waiting'
  | 'blocked'
  | 'review'
  | 'completed'
  | 'cancelled';
export type WorkItemPriority = 'low' | 'normal' | 'high' | 'critical';
export type WorkItemActorType = 'user' | 'system' | 'integration';

export interface PartyReference {
  id?: string;
  displayName: string;
}

export interface WorkItemContextEntry {
  id: string;
  type: 'note' | 'source' | 'reference' | 'decision';
  text?: string;
  url?: string;
  createdAt: string;
  actorType: WorkItemActorType;
  actorName?: string;
}

export interface ExternalReference {
  type: string;
  label: string;
  id?: string;
  url: string;
}

export interface WorkItemEvent {
  id: string;
  sequence?: number;
  at: string;
  actorType: WorkItemActorType;
  actorName?: string;
  action: string;
  summary: string;
  changes?: Array<{
    field: string;
    from?: unknown;
    to?: unknown;
  }>;
}

/**
 * Stable provider-neutral contract exposed by the application layer.
 *
 * The existing Task interface remains a compatibility/persistence format while
 * the UI and integrations are migrated. New product concepts belong here rather
 * than in the legacy Task aggregate.
 */
export interface WorkItem {
  schemaVersion: typeof WORK_ITEM_SCHEMA_VERSION;
  id: string;
  kind: WorkItemKind;
  obligationMode?: ObligationMode;
  legacyObligationKind?: TaskObligationKind;
  title: string;
  description?: string;
  expectedOutcome?: string;
  status: WorkItemStatus;
  priority: WorkItemPriority;
  owner?: PartyReference;
  accountableTo?: PartyReference;
  areaId?: string;
  parentId?: string;
  startAt?: string;
  dueAt?: string;
  completedAt?: string;
  nextReviewAt?: string;
  blockerReason?: string;
  source: string;
  sourceUrl?: string;
  planningBucket?: PlanningBucket;
  estimateMinutes?: number;
  externalReferences: ExternalReference[];
  tags: string[];
  context: WorkItemContextEntry[];
  createdAt: string;
  updatedAt: string;
  revision: number;
  archivedAt?: string;
  history: WorkItemEvent[];
  metadata?: Record<string, unknown>;
}

const LEGACY_STATUS_MAP: Record<TaskStatus, WorkItemStatus> = {
  new: 'planned',
  analyzed: 'ready',
  'in-progress': 'in_progress',
  'ready-for-review': 'review',
  done: 'completed',
  blocked: 'blocked',
};

export function workItemKindFromLegacy(kind: TaskObligationKind | undefined): WorkItemKind {
  return !kind || kind === 'task' ? 'task' : 'obligation';
}

export function workItemStatusFromTask(task: Task): WorkItemStatus {
  if (task.waitingState) return 'waiting';
  if (task.attentionState === 'pr-comments') return 'review';
  return LEGACY_STATUS_MAP[task.status];
}

export function workItemPriorityFromTask(task: Task): WorkItemPriority {
  const score = task.priorityScore ?? 0;
  if (score >= 85) return 'critical';
  if (score >= 65) return 'high';
  if (score > 0 && score < 30) return 'low';
  return 'normal';
}

/** Lossless compatibility projection for active UI reads during migration. */
export function taskToWorkItem(task: Task): WorkItem {
  const createdAt = task.createdAt ?? task.receivedAt;
  const kind = workItemKindFromLegacy(task.obligationKind);
  const context: WorkItemContextEntry[] = [];

  if (task.notes?.trim()) {
    context.push({
      id: `legacy-note-${task.id}`,
      type: 'note',
      text: task.notes,
      createdAt: task.updatedAt ?? createdAt,
      actorType: 'user',
    });
  }
  if (task.originalMessage?.trim()) {
    context.push({
      id: `legacy-source-${task.id}`,
      type: 'source',
      text: task.originalMessage,
      url: task.sourceUrl,
      createdAt: task.receivedAt,
      actorType: 'integration',
    });
  }

  const projected: WorkItem = {
    schemaVersion: WORK_ITEM_SCHEMA_VERSION,
    id: task.id,
    kind,
    ...(kind === 'obligation'
      ? {
          obligationMode: task.obligationKind === 'responsibility' ? 'ongoing' as const : 'one_off' as const,
          legacyObligationKind: task.obligationKind,
        }
      : {}),
    title: task.title,
    ...(task.description ? { description: task.description } : {}),
    status: workItemStatusFromTask(task),
    priority: workItemPriorityFromTask(task),
    ...(task.responsibleParty
      ? { owner: { displayName: task.responsibleParty } }
      : {}),
    ...(task.accountableTo
      ? { accountableTo: { displayName: task.accountableTo } }
      : {}),
    ...(task.customerId ? { areaId: task.customerId } : {}),
    ...(task.dueAt ? { dueAt: task.dueAt } : {}),
    ...(task.completedAt ? { completedAt: task.completedAt } : {}),
    source: task.source,
    ...(task.sourceUrl ? { sourceUrl: task.sourceUrl } : {}),
    ...(task.planningBucket ? { planningBucket: task.planningBucket } : {}),
    externalReferences: [
      ...(task.devopsTaskUrl ? [{ type: 'azure_devops_work_item', label: 'Azure DevOps', url: task.devopsTaskUrl }] : []),
      ...(task.ticketUrl ? [{ type: 'ticket', label: 'Helpdesk ticket', url: task.ticketUrl }] : []),
    ],
    tags: [],
    context,
    createdAt,
    updatedAt: task.updatedAt ?? createdAt,
    revision: Math.max(1, task.revision ?? 1),
    ...(task.archivedAt ? { archivedAt: task.archivedAt } : {}),
    history: (task.history ?? []).map((entry) => ({
      ...entry,
      changes: entry.changes?.map((change) => ({ ...change })),
    })),
  };
  const embedded = (task as Task & { _canonicalWorkItem?: WorkItem })._canonicalWorkItem;
  if (!embedded || embedded.id !== task.id || embedded.schemaVersion !== WORK_ITEM_SCHEMA_VERSION) {
    return projected;
  }
  const projectedContextIds = new Set(projected.context.map((entry) => entry.id));
  return {
    ...embedded,
    ...projected,
    expectedOutcome: embedded.expectedOutcome,
    parentId: embedded.parentId,
    startAt: embedded.startAt,
    nextReviewAt: embedded.nextReviewAt,
    blockerReason: embedded.blockerReason,
    tags: embedded.tags,
    context: [
      ...embedded.context.filter((entry) => !projectedContextIds.has(entry.id)),
      ...projected.context,
    ],
    metadata: embedded.metadata,
  };
}
